SSA/Projects/InstCombine/Base.lean

```diff
@@ -102,7 +118,22 @@ inductive MOp.BinaryOp : Type
   | sdiv
   | udiv
 deriving Repr, DecidableEq, Inhabited
-
+def aux : MOp.BinaryOp → ℕ → Lean.Format
+  | .and => reprPrec g.and
+  | .or => reprPrec g.or
+  | .xor => reprPrec g.xor
+  | .shl => reprPrec g.shl
+  | .lshr => reprPrec g.lshr
+  | .ashr => reprPrec g.ashr
+  | .urem => reprPrec g.urem
+  | .srem => reprPrec g.srem
+  | .add _ => reprPrec g.add
+  | .mul => reprPrec g.mul
+  | .sub => reprPrec g.sub
+  | .sdiv => reprPrec g.sdiv
+  | .udiv => reprPrec g.udiv
+instance : Repr (MOp.BinaryOp) where
+  reprPrec op w := ((toString (aux op w)).replace "InstCombine.g" "InstCombine.MOp.BinaryOp").replace "false" ""
```

Comment:
This code is quite obscure, and it's not super clear to me, but I guess the purpose of this is to "hide" the default `false` values? Why do we need those to be hidden?

I guess the purpose of `g` is to use it's automatically derived `Repr` instance here? This is feels like a very roundabout way of achieving that goal. 

Consider something like the following: 
```lean
def repr : ...
  | .and => "and"
  | .or => "or"
  ...
  | .add flags =>
    let \<nsw, nuw \> := flags
    if nsw = false \land nuw = false then
      "add"
    else
      "add " ++ repr flags
```
This is much more readable, and means you can get rid of `g`, so it should be significantly simpler as well

What 

This is a very r

I guess the purpose of introducing the `g` type is to be able to

Reply:
> ```
> /--
> The reason that I am using the admittedly hacky and ad-hoc method is that I want to preserve the guard_msgs > statements, otherwise the build will fail.
> But the default Repr instance has some fancy behavior where depending on the indentation it will sometimes wrap in parentheses.
> I think the only way to replicate this behavior is to have another class and piggy-back off its default Repr class
> -/
```

Thanks for writing this comment to explain your rationale! You are right that this solution deserves some comment in the code, but first this would fit better here in the code review --- also, as an aside, code comments are best not written in the first person, since a future reader might not immediately know who actually wrote that comment.

More concretely, the fancy behaviour you refer to is simply behaviour of [Std.Format](https://leanprover-community.github.io/mathlib4_docs/Init/Data/Format/Basic.html#Std.Format).  You can have a look at the [derive handler of Repr](https://github.com/leanprover/lean4/blob/c375e19f6b656fcd594cdca3a38b8578634df8cd/src/Lean/Elab/Deriving/Repr.lean#L82-L86) to see what code it generates; in particular the line: `Repr.addAppParen (Format.group (Format.nest (if prec >= max_prec then 1 else 2) ($rhs:term))) prec)` is where the parenthesization happens.

We could write similar logic by hand. See for example:
```
open Std (Format) in
def reprWithoutFlags (op : MOp.BinaryOp) (prec : Nat) : Format :=
  let op := match op with
    | .and    => "and"
    | .or     => "or"
    | .xor    => "xor"
    | .shl    => "shl"
    | .lshr   => "lshr"
    | .ashr   => "ashr"
    | .urem   => "urem"
    | .srem   => "srem"
    | .add _  => "add"
    | .mul    => "mul"
    | .sub    => "sub"
    | .sdiv   => "sdiv"
    | .udiv   => "udiv"
  Repr.addAppParen (Format.group (Format.nest
    (if prec >= max_prec then 1 else 2) f!"InstCombine.MOp.BinaryOp.{op}"))
    prec
```

Finally, and the reason I do consider this one a blocker for merging: your current `Repr` instance completely ignores flags, all four of the following lines print exactly the same (`InstCombine.MOp.BinaryOp.add`)
```
#eval repr (MOp.BinaryOp.add ⟨false, false⟩)
#eval repr (MOp.BinaryOp.add ⟨false, true⟩)
#eval repr (MOp.BinaryOp.add ⟨true, true⟩)
#eval repr (MOp.BinaryOp.add ⟨true, false⟩)
```